### ISSUE TITLE
Asynchronous log sinks with fern?

### ISSUE THREAD

**Issue by mersinvald** [daboross/fern#15]
I'm planning to use fern in the asynchronous app and non-blocking logging is vital for us.

How can I setup fern to implement async logging? Can it write logs in another thread or at least use mpsc::channel as an output?

**Comment by daboross** [comment]
Hi!

Thank you for bringing this up - this isn't currently possible in `fern`, but it is something I've been wanting to add.

In the ideal scenario, how much control would you want over how `fern` does this?

It would be fairly easy to add a `.through_thread()` option which just had fern spawn up a single logging thread at that level. Outputting to a `std::sync::mpsc::Sender<String>` would also be a fairly easy addition.

**Comment by daboross** [comment]
Integrating directly with `tokio` could also be a possibility, but I imagine it would be a more complicated process to make ergonomic than simply having existing loggers operate in a single separate thread.

**Comment by mersinvald** [comment]
@daboross yes, I believe that a thread would be a perfect solution here, the simplest at least.

Tokio integration could work too, I imagine that something like this wouldn't add up any complication to `fern` usage, for ones who works with Tokio anyway:
```rust
let mut core = Core::new()?;
let handle = core.handle();

fern::Dispatch::new()
    .with_tokio(handle)
    .chain(file)
    .apply()?;

// Spawning an async write to file task onto a Tokio Core with the Handle
info!("async!") 
```
But I believe it will be non-trivial to implement alongside a synchronous logging.

And lastly outputting to mpsc::Sender option is good with it's flexibility and simplicity: it would be straightforward to just send our logs to any logging daemon from a separate thread that receives logs from fern via mpsc, for example.

**Comment by mersinvald** [comment]
Also the mpsc::Sender option is pretty simple and I might be able to implement it as an fern::Output myself, if you don't mind and send you a PR?

**Comment by daboross** [comment]
Not at all, that would be great!

I'm not sure if we want any extra information sent alongside the record to mpsc::Sender - it would be simplest just to send a fully formatted String with the output result.

I'll start work on getting a threaded option built-in as well. The main concern I have is reconstructing the `log::LogRecord` in the other thread - the `log` crate doesn't currently expose any way to construct records other than the `log!()` macro. We can definitely get something working before that's added to `log` - but the filename, line number and column number will need to be inaccurate.

**Comment by mersinvald** [comment]
Oops, didn't realize that our issues will reference yours.

By the way, I'm delegating the task to my "colleague-rust-beginner" @gingerDevilish, so do expect a PR from her 😃

**Comment by daboross** [comment]
It's all good, connecting things can't be too bad.

I'll be open to a future PR! If any more information is needed on `fern`'s structure or other bits, I'm definitely open to questions as well.

**Comment by daboross** [comment]
Support for `std::sync::mpsc::Sender` is now added as of version 0.4.3.

I'm still aiming to add direct threading support into `fern`, but it may have to wait until after more changes to the `log` crate.

Thanks @gingerDevilish for what we have now!

**Comment by rcashie** [comment]
Stalked @gingerDevilish and his repos for an example:
https://github.com/DevHeap/circles-server/blob/e1151fa5c5fde123669fab3c5858fac5aaa2de6f/services/positions-api/src/main.rs#L43

Would be great if such an example was in this repo.

**Comment by vorner** [comment]
Might be a bit related to this ‒ I have my own async logging implementation. I was considering to create it as a general log wrapper, but I wanted to integrate a thread name into it too, which wasn't possible.

https://docs.rs/spirit-log/0.2.2/spirit_log/background/index.html
